vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/resource/DocumentV1ApiHandler.java

```diff
@@ -243,24 +253,44 @@ public void handleTimeout(Request request, ResponseHandler responseHandler) {
 
     @Override
     public void destroy() {
+        Instant doom = clock.instant().plus(Duration.ofSeconds(30));
+
+        // This blocks until all visitors are done. These, in turn, may require the asyncSession to be alive
+        // to be able to run, as well as dispatch of operations against it, which is done by visitDispatcher.
+        visits.values().forEach(VisitorSession::destroy);
+
+        // Shut down both dispatchers, so only we empty the queues of outstanding operations, and can be sure they're empty.
         dispatcher.shutdown();
-        Instant doom = clock.instant().plus(Duration.ofSeconds(20));
-        while ( ! operations.isEmpty() && clock.instant().isBefore(doom))
+        visitDispatcher.shutdown();
+        while ( ! (operations.isEmpty() && visitOperations.isEmpty()) && clock.instant().isBefore(doom)) {
             dispatchEnqueued();
+            dispatchVisitEnqueued();
+        }
 
         if ( ! operations.isEmpty())
             log.log(WARNING, "Failed to empty request queue before shutdown timeout — " + operations.size() + " requests left");
 
-        asyncSession.destroy();
-        visits.values().forEach(VisitorSession::destroy);
+        if ( ! visitOperations.isEmpty())
+            log.log(WARNING, "Failed to empty visitor operations queue before shutdown timeout — " + operations.size() + " operations left");
 
         try {
+            while (outstanding.get() > 0 && clock.instant().isBefore(doom))
```

Comment:
Messagebus guarantees a response to each successfully sent message. This counter is incremented whenever an operation is accepted by the document access, which corresponds to this.